source/MaterialXCore/Interface.cpp

```diff
@@ -329,8 +329,15 @@ OutputPtr Input::getConnectedOutput() const
     // Look for output on a node
     else if (hasNodeName())
     {
+        const string& nodeName = getNodeName();
         ConstGraphElementPtr graph = getAncestorOfType<GraphElement>();
-        NodePtr node = graph ? graph->getNode(getNodeName()) : nullptr;
+        NodePtr node = graph ? graph->getNode(nodeName) : nullptr;
```

Comment:
For a given nodegraph input check at the parent nodegraph level for a connection.

Reply:
I think there is a risk of ambiguity here, if a node outside the graph has the same name as a node inside the graph. It's a possible setup since they have two different naming scopes.

So if a nodegraph input connects to a node "foo1" outside the graph (nodename="foo1"), but there is also a node inside the graph named "foo1", the node inside the graph will be used, since both scopes are checked and inside nodes takes precedence. This would naturally be invalid since a node inside the graph is not connectable to a nodegraph input.

I think we have to handle graph inputs and node inputs separately in this traversal, since graph inputs can only connect to elements _outside_ the graph scope, and node inputs can only connected to elements in the same scope as the node.

Reply:
I was trying to make this as reusable as possible. I'll need to take a closer look but getAncestorOfType<>() on an input in a graph is what being handled here which requires traversing twice as the first returned value is the graph itself and not it's parent.  In both the nodegraph and node case it should be looking for a parent.

Reply:
I'm mainly thinking of name collisions with nodes inside vs outside the graph scope. As in this example:
```
<constant name="foo1" type="color3">
    <parameter name="value" type="color3" value="1, 0, 0"/>
</constant>
<nodegraph name="graph1" >
    <input name="in1" type="color3" nodename="foo1" />
    <input name="in2" type="color3" value="0, 1, 0" />
    <constant name="foo1" type="float">
        <parameter name="value" type="float" value="0.5"/>
    </constant>
    <mix name="mix1" type="color3">
      <input name="bg" type="color3" interfacename="in1" />
      <input name="fg" type="color3" interfacename="in1" />
      <input name="mix" type="float" nodename="foo1" />
    </mix>
    <output name="out" type="color3" nodename="mix1" />
</nodegraph>
```
If the internal scope is always searched first, in this case the internal node "foo1" would be returned as being connected to the input "in1", when in fact it is the external node "foo1" that is connected.

A test case like this could be added to make sure this is handled correctly.

Reply:
@bernardkwok @niklasharrysson As discussed in a separate thread, I think it would be valuable to merge the latest work from the v1.38 branch of MaterialX before proceeding too far with these new improvements, as there's a good deal of code that's been removed from Interface.cpp.

Reply:
I've done the merge so this logic still needs to be resolved :). 
@niklasharrysson, looking at the same level will give the wrong result for "in1".  So I guess the branching logic would need to be:
if (input is a child of a nodegraph)
{ 
     look for referenced node/graph from parent of nodegraph.
}
else
{ 
    look for referenced node/graph from parent of node.
}

This basically means to always look up 2 levels from the input (if they exist). 

We don't have the concept of a input which is not a child of a node or nodegraph (i.e. at the Document level) so this does not need to be considered.